### DISCUSSION TITLE
Transporting the setup for other controllers

### DISCUSSION THREAD

**Discussion by Astronokum** [raphaelquast/beatstep#84] [Q&A]
Hey,
im experiencing a near-flawless functionality when using a beatstep pro with the beatstep_q setup.
unfortunately there seems to be one litte issue with the encoder modes.
all the controls in ableton can only be "turned up" with the knobs.
i only got low-tier coding experience (as well as mid-tier midi insight) but i assume its a quick fix... my guess was that the correct behaviour should be "absolute" instead of "smooth relative two".
before messing up the script i thought id ask you guys for hints or ideas... 
thank you very much for all the work and time spent for making this great tool!
Julian

**Comment by raphaelquast** [comment]
Hey, 
Thanks! Awesome to hear that the script seems to be compatible with the beatstep pro! 
I was wondering if that's possible for quite some time, but since I don't own one, I never had the chance to try.

If the encoder-mode does the trick, it should indeed only require minor changes.  
Fortunately I implemented all MIDI Sysex messages I could find when I started to experiment with the script ([here](https://github.com/raphaelquast/beatstep/blob/e3948866e6d72713f3038a7e3cc773a3886db5a3/QSetup.py#L29C1-L45C1)) ... and for the encoder-modes there's this function:
```python
def set_E_behaviour(ID, val):
    """
    ID: int
        The encoder number (1-16)
    val: int
        Set encoder behaviour
        (0=absolute, 1-3=relative mode 1-3)
    """"
```

It is called on startup (and on layer-change etc.) to make sure all encoders are always in `relative mode 2` right now.

To switch to absolute, all you have to do is to change **all occurances** of `self._send_midi(self.QS.set_E_behaviour(i, 2))` to `self._send_midi(self.QS.set_E_behaviour(i, 0))` in [Beatstep_Q.py](https://github.com/raphaelquast/beatstep/blob/master/BeatStep_Q.py)!

If you need to change the mode of the "transpose" encoder, change `self._send_midi(self.QS.set_E_behaviour("transpose", 0))`.

> Note that you need to restart Ableton if you change something in a MIDI remote script because they are cached on startup!
